src/lib/Builder.hs

```diff
@@ -740,9 +740,14 @@ buildEffLam
 buildEffLam rws hint ty body = do
   eff <- getAllowedEffects
   buildLam noHint PlainArrow TyKind Pure \h -> do
-    let eff' = extendEffect (RWSEffect rws (Just h)) (sink eff)
-    buildLam hint PlainArrow (RefTy (Var h) (sink ty)) eff' \ref ->
-      body (sink h) ref
+    let ty' = RefTy (Var h) (sink ty)
+    withFreshBinder hint (LamBinding PlainArrow ty') \b -> do
+      let ref = binderName b
+      h' <- sinkM h
+      let eff' = extendEffect (RWSEffect rws (Just h')) (sink eff)
+      body' <- withAllowedEffects eff' $ buildBlock $ body (sink h) $ sink ref
+      effs <- effectsE body'
+      return $ Lam $ LamExpr (LamBinder b ty' PlainArrow effs) body'
```

Comment:
This change is a little unclear to me. Why did we even have to do it? Is there some subtle interplay between your patches and `getAllowedEffects`?

Reply:
Would this be fixed if only we adjusted `buildLam`?

Reply:
Why do this at all?  This way, the effects associated with an effect lambda are more precise than before.  The difference is that this implementation extracts the effects from the block and uses them in the `LamBinder`, whereas `buildLam` puts the input effects `eff'` in the `LamBinder`.

Why do this here and not in `buildLam`?  Dougal and I went back and forth on this a bit.  The issue, as I currently understand it, is that an `EffectRow` can mean one of three different things: it can mean the set of effects permitted in some context (which is what `getAllowedEffects` produces); it can mean the set of effects actually used by some code (which is what the annotations on `Block` and `Case` now mean); and it can be part of the type of a function, where it serves both purposes (limiting what's allowed in the body, and indicating what's demanded by calls).

So, at the level of `buildLam`, we have to treat the input effect row as defining the type of the function be are building, which in general we are not free to modify.  However, at the more specific `buildEffLam`, we know the function is about to be consumed by an effect discharge primitive (I think?), so we actually are free to produce a function with a more specific type.

Historical note: I actually tried it the other way first, always tagging every `LamExpr` with the effects consumed by the `Block`, but hit a test failure I didn't understand.  Dougal gave me the above explanation about function types, and we decided to back off to this more limited change.

Added a brief comment to the code to allude to this explanation.